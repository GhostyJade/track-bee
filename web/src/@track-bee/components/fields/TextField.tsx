import { withFormsy, FormsyInjectedProps } from 'formsy-react';
import { TextField as MUITextField } from '@mui/material';

type TextFieldPropsBase = {
    name: string;
    label?: string;
    value?: string | number;
    type?: 'text' | 'password' | 'number';
    onBlur?: (event: React.FocusEvent<HTMLInputElement>) => void;
    onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
    required?: boolean;
    helperText?: string | JSX.Element | null;
};

export interface TextFieldProps extends Omit<TextFieldPropsBase, 'onChange' | 'onBlur'> {
    sendValueChange(name: string, value: string): void;
}

function TextFieldInternal({ name, label, value, type, onBlur, onChange, required, helperText }: TextFieldPropsBase) {
    return (
        <MUITextField
            name={name}
            label={label}
            value={value}
            type={type}
            onBlur={onBlur}
            onChange={onChange}
            required={required}
            helperText={helperText}
        />
    );
}

export default function TextField({ sendValueChange, ...props }: TextFieldProps): JSX.Element {
    const onBlur = (event: React.FocusEvent<HTMLInputElement>) => {
        sendValueChange(event.target.name, event.target.value);
    };

    return <TextFieldInternal {...props} onBlur={onBlur} />;
}

type FormsyTextFieldProps = TextFieldProps & FormsyInjectedProps<string | number>;

function FormsyTextFieldInternal({ sendValueChange, errorMessage, ...props }: FormsyTextFieldProps) {
    const onBlur = (event: React.FocusEvent<HTMLInputElement>) => {
        sendValueChange(event.target.name, event.target.value);
    };

    const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        props.setValue(event.target.value);
    };

    return (
        <TextFieldInternal
            {...props}
            helperText={errorMessage ? <p>{errorMessage}</p> : null}
            onBlur={onBlur}
            onChange={onChange}
        />
    );
}

export const FormsyTextField = withFormsy(FormsyTextFieldInternal);
