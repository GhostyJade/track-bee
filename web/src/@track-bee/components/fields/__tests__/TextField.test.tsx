/* eslint-disable @typescript-eslint/no-explicit-any */

import React from 'react';
import renderer from 'react-test-renderer';

import TextField from '../TextField';

beforeEach(() => {
    const setStateMock = jest.fn();
    const useStateMock: any = (useState: any) => [useState, setStateMock];

    jest.spyOn(React, 'useState').mockImplementation(useStateMock);
});

describe('TextField', () => {
    const [value, setValue] = React.useState('');
    const component = renderer.create(<TextField name='test' value={value} sendValueChange={(_, v) => setValue(v)} />);

    it('should render without content', () => {
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should write content to the component', () => {
        const tree = component.toTree();
        renderer.act(() => {
            tree?.props.sendValueChange('', 'aaaaa');
        });
        expect(tree?.props.value).toEqual('aaaaa');
        expect(component.toJSON()).toMatchSnapshot();
    });
});
