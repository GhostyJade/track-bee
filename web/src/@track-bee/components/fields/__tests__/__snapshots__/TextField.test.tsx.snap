// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField should render without content 1`] = `
<div
  className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
>
  <div
    className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
      disabled={false}
      id=":r0:"
      name="test"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={false}
      type="text"
      value=""
    />
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-hdw1oc"
      >
        <span
          className="notranslate"
        >
          â€‹
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
